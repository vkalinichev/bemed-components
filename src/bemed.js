import React from 'react'
import tags from './tags'
import addModificators from './addModificators'

function BemedComponent(tagName, className, {mod, ...rest}) {
  return React.createElement(
    tagName,
    {
      ...rest,
      className: addModificators(className, mod)
    })
}

/**
 * Creates Block context and returns Block-generator
 * @param blockName
 * @returns {b} Generator of Blocks and Elements
 */
function bemed(blockName) {

  /**
   * Generator of Blocks and Elements
   * @param {string} tagName - (optional) custom tag for Block or Element
   * @param {string} elementName
   * @returns {Component}
   * @returns {function} Genetator with custom Component
   */
  function b(tagName, elementName) {

    if (typeof tagName === 'function') {
      return b.bind(null, tagName, elementName)
    }

    if (elementName === undefined && tags.indexOf(tagName) === -1 ) {
      elementName = tagName
      tagName = 'div'
    }

    const isBlock = elementName === undefined
    const className = isBlock ? blockName : `${ blockName }__${ elementName }`

    const Component = BemedComponent.bind(null, tagName, className)
    Component.displayName = className

    return Component
  }

  for (let i = 0; i < tags.length; i++) {
    b[tags[i]] = b.bind(null, tags[i])
  }

  return b
}

export default bemed
