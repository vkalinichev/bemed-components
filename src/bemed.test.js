import React from 'react'
import bemed from './bemed'
import renderer from 'react-test-renderer'

describe('Block', () => {

  it('should create block', () => {
    const Block = bemed('block-name')()
    const component = renderer.create(<Block/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should create block with custom tag (SPAN)', () => {
    const Block = bemed('block-name').span()
    const component = renderer.create(<Block/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

})

describe('Element', () => {

  it('should create element', () => {
    const b = bemed('block-name')
    const Element = b('element')
    const component = renderer.create(<Element/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should create element with custom tag (SPAN)', () => {
    const b = bemed('block-name')
    const Element = b.span('element')
    const component = renderer.create(<Element/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

})

describe('Modifier', () => {

  it('should add string modifier for block', () => {
    const b = bemed('block-name')
    const Element = b()
    const component = renderer.create(<Element mod="cool"/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add few string modifiers for block', () => {
    const b = bemed('block-name')
    const Element = b()
    const component = renderer.create(<Element mod="cool feel"/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add truly modifier for block', () => {
    const b = bemed('block-name')
    const Element = b()
    const component = renderer.create(<Element mod={{cool: true}}/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should not add falsy modifier for block', () => {
    const b = bemed('block-name')
    const Element = b()
    const component = renderer.create(<Element mod={{cool: false}}/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add valued modifier for block', () => {
    const b = bemed('block-name')
    const Element = b()
    const component = renderer.create(<Element mod={{cool: 'cool'}}/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add string modifier for element', () => {
    const b = bemed('block-name')
    const Element = b('element')
    const component = renderer.create(<Element mod="cool"/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add few string modifiers for element', () => {
    const b = bemed('block-name')
    const Element = b('element')
    const component = renderer.create(<Element mod="cool feel"/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add truly modifier for element', () => {
    const b = bemed('block-name')
    const Element = b('element')
    const component = renderer.create(<Element mod={{cool: true}}/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should not add falsy modifier for element', () => {
    const b = bemed('block-name')
    const Element = b('element')
    const component = renderer.create(<Element mod={{cool: false}}/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should add valued modifier for element', () => {
    const b = bemed('block-name')
    const Element = b.span('element')
    const component = renderer.create(<Element mod={{cool: 'cool'}}/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('Early configuration', () => {

    it('should add truly modifier for element', () => {
      const b = bemed('block-name')
      const Element = b('element', { mod: ['cool']})
      const component = renderer.create(<Element cool/>)
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should add few truly modifiers for element', () => {
      const b = bemed('block-name')
      const Element = b('element', { mod: ['cool', 'feel']})
      const component = renderer.create(<Element cool feel/>)
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should not add falsy modifier for element', () => {
      const b = bemed('block-name')
      const Element = b('element', { mod: ['cool']})
      const component = renderer.create(<Element cool={false}/>)
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should add valued modifier for element', () => {
      const b = bemed('block-name')
      const Element = b.span('element', { mod: ['cool']})
      const component = renderer.create(<Element cool='yes'/>)
      expect(component.toJSON()).toMatchSnapshot()
    })

  })

})

describe('Custom Components', () => {

  it('should create block', () => {
    const b = bemed('block-name')
    const FancyComponent = (props) => (
      <fancy-element {...props}>
        FancyComponent
      </fancy-element>
    )
    const Element = b(FancyComponent)
    const component = renderer.create(<Element/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should create element', () => {
    const b = bemed('block-name')
    const FancyComponent = (props) => (
      <fancy-element {...props}>
        FancyComponent
      </fancy-element>
    )
    const Element = b(FancyComponent, 'element')
    const component = renderer.create(<Element/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

})

describe('Miccelaneous', () => {

  it('should provide global classNames for block', () => {
    const Block = bemed('block-name')()
    const component = renderer.create(<Block className="global another-global"/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should provide global classNames for element', () => {
    const b = bemed('block-name')
    const Element = b('element')
    const component = renderer.create(<Element className="global another-global"/>)
    expect(component.toJSON()).toMatchSnapshot()
  })

})
