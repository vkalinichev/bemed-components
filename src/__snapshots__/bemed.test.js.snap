// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block should create block 1`] = `
<div
  className="block-name"
/>
`;

exports[`Block should create block with custom tag (SPAN) 1`] = `
<span
  className="block-name"
/>
`;

exports[`Custom Components should create block 1`] = `
<fancy-element
  className="block-name"
>
  FancyComponent
</fancy-element>
`;

exports[`Custom Components should create element 1`] = `
<fancy-element
  className="block-name__element"
>
  FancyComponent
</fancy-element>
`;

exports[`Element should create element 1`] = `
<div
  className="block-name__element"
/>
`;

exports[`Element should create element with custom tag (SPAN) 1`] = `
<span
  className="block-name__element"
/>
`;

exports[`Miccelaneous should provide global classNames for block 1`] = `
<div
  className="block-name global another-global"
/>
`;

exports[`Miccelaneous should provide global classNames for element 1`] = `
<div
  className="block-name__element global another-global"
/>
`;

exports[`Modifier Early configuration should add few truly modifiers for element 1`] = `
<div
  className="block-name__element  block-name__element_cool block-name__element_feel"
/>
`;

exports[`Modifier Early configuration should add truly modifier for element 1`] = `
<div
  className="block-name__element  block-name__element_cool"
/>
`;

exports[`Modifier Early configuration should add valued modifier for element 1`] = `
<span
  className="block-name__element  block-name__element_cool_yes"
/>
`;

exports[`Modifier Early configuration should not add falsy modifier for element 1`] = `
<div
  className="block-name__element"
/>
`;

exports[`Modifier should add few string modifiers for block 1`] = `
<div
  className="block-name block-name_cool block-name_feel"
/>
`;

exports[`Modifier should add few string modifiers for element 1`] = `
<div
  className="block-name__element block-name__element_cool block-name__element_feel"
/>
`;

exports[`Modifier should add string modifier for block 1`] = `
<div
  className="block-name block-name_cool"
/>
`;

exports[`Modifier should add string modifier for element 1`] = `
<div
  className="block-name__element block-name__element_cool"
/>
`;

exports[`Modifier should add truly modifier for block 1`] = `
<div
  className="block-name  block-name_cool"
/>
`;

exports[`Modifier should add truly modifier for element 1`] = `
<div
  className="block-name__element  block-name__element_cool"
/>
`;

exports[`Modifier should add valued modifier for block 1`] = `
<div
  className="block-name  block-name_cool_cool"
/>
`;

exports[`Modifier should add valued modifier for element 1`] = `
<span
  className="block-name__element  block-name__element_cool_cool"
/>
`;

exports[`Modifier should not add falsy modifier for block 1`] = `
<div
  className="block-name"
/>
`;

exports[`Modifier should not add falsy modifier for element 1`] = `
<div
  className="block-name__element"
/>
`;
